### ISSUE TITLE
Change delimiter in OwnerOperator

### ISSUE THREAD

**Issue by tlageman-epa** [US-EPA-CAMD/PI-1#4888]
### Definition of Ready
- [ ] Does this meet the INVEST criteria?
- [ ] Does this have a clearly defined user?
- [ ] Does this have an estimate?
- [ ] Does this have acceptance criteria?
- [ ] Does this have sufficient context for the specific functionality being developed (design, business rules, etc.)
- [ ] Does this have a dev link been provided in the design?
- [ ] Is this traceable to a feature, epic, or PI?

## User Story
As a 
I want
So that

## Acceptance Criteria
Given
When
Then

### Does this Story require 508 Testing?
- [ ] Implementing a new authoring tool that will produce code / content for the application
- [ ] Implementing a 3rd party widget that will be part of the application
- [ ] Adding new / updating custom functionality 
- [ ] Adding new / updating a page template
- [ ] Adding new / replacing existing functionality (unless the replacement has already passed 508 testing)
- [ ] Full Site Review prior to a scheduled EPA 508 review
- [ ] Planning a video that will appear on the application
- [ ] Producing a video that will appear on the application
- [ ] Producing Training materials
- [ ] Producing documents

## Context
3/7/2023 (tlageman) - Current delimiter is a comma (,). This makes it impossible to parse out different Owner Operators from the outputted data. Previously, the old system had a "< br >" as the delimiter. Recommend using a "|" as we have for other parts of the new EASEY system.

This affects Facility/Unit Attributes, and all allowance and compliance data sub types and their streaming services counterparts.

## Testing
Test Case:
Tester:

### Definition of Done
- [ ] Have all Acceptance Criteria been agreed to and validated by the scrum team?
- [ ] Has the code been unit tested and peer reviewed?
- [ ] Have the functional tests been executed?
- [ ] Have all defects been reviewed and dispositioned, resolved, or deferred?
- [ ] Is the user story ready to be deployed to the test/staging environments?

**Comment by shahbazkhan1999** [comment]
fixed the test cases and also update the node version along with delimiter changes
https://github.com/US-EPA-CAMD/easey-streaming-services/pull/94

**Comment by ergjustin** [comment]
@shahbazkhan1999  Build is failed. Need to fix build error (update node version in github->workflow-> .yml files)

**Comment by ergjustin** [comment]
@shahbazkhan1999 
Can you please add some local env testing screenshot.

**Comment by shahbazkhan1999** [comment]
1. Bottom of our app's main page first click on the button "**Start your data query**"
 (https://campd-stg.app.cloud.gov/)
2. Then move to the next page where we have data type select the data type "**compliance**" and also select the sub-data type "**Allowance based**"
3.  After the selection of both click on the apply button and next apply each filter one by one in all filter '|' delimiter is applied check-in data table owner/operator column this delimiter is applied 
4.  Also download the CSV and check the owner/operator column. 
5.  I added some attachments for both CSV and Data Table views. Check these Attachments links below.

[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMWQzQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--fb9fb22e6c7da7c7b816e8e291095ce81cbb6e2d/facility.png)

[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMWgzQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--fab741b69b8288a43032bd490494cb2120849832/facility%20csv.png)
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMXAzQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--b7a522a6879769e1ddf13c9ed87f5cdd6f2c17a3/ownerOperator-stage.png)
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMXQzQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--8c74cc8ab7008752b4fda237f10d105b8491a326/ownerOperator%20Csv.png)

**Comment by tlageman-epa** [comment]
A few corrections @shahbazkhan1999 : 
- account info - data preview is missing | and shows , still. The download looks good though.
- allowance holdings - data preview has the | but the download is still using the ,
- allowance transactions - data preview and download are still using the , Keep in mind that transactions has Owner (Transferor) and Owner (Transferee) fields that both need the | instead of the ,
- compliance emissions - data preview and download are still using the ,

**Comment by ergjustin** [comment]
@shahbazkhan1999 
Allowance-> Account Information Download Not same as display
[Account Information Download](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMmFVQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--6e945408be0d099e64a271f4da1d775468052a20/test.PNG)

**Comment by alangmaid** [comment]
@shahbazkhan1999 examples of improper "|" in the csv remain.
Compliance - Allowance Based. When the "Owner/Operator" data contain a "," the CSV incorrectly display a "|". The HTML correctly displays  ","  [Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBM2FVQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--a9b61c53c19c2be9fad9b6a7c1bdc6b6d72a89b2/image.png)
I believe this bug occurs for **ALL** CSV files that include the Owner/Operator field including Facility data type. I.e., when the "Owner/Operator" name includes a "," it is incorrectly displaying as a "|". E.g., "PowerSouth Energy Cooperative |  Inc. (Operator)" should display as "PowerSouth Energy Cooperative, Inc. (Operator)"

**Comment by alangmaid** [comment]
1) Allowance - Transactions preview AND CSV display a "," between names in the "Owner (Transferee)" and "Owner (Transferor)" fields instead of "|"
For example, 
AEP Generation Resources, Inc.,Buckeye Power, Inc.
should display as
AEP Generation Resources, Inc. | Buckeye Power, Inc.

2) Compliance - Allowance based preview display a "|" in "Owner/Operator" when it should display a ",".
For example, search on program = CSSO2G1 and preview. Ctrl-F "Springfield"
[Preview screen](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNldtQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--3afcc6dcf0764bd5e86380452a6a330d48fd3c1c/image.png)

**Comment by tlageman-epa** [comment]
ownerOperator is found in the following APIs/endpoints and should have the | delimiter applied between each owner/operator value in a record e.g., "Georgia Power Company (Owner)|Southern Company (Owner)|Alabama Power Company (Operator)" which should be returned in the API call, CAMPD UI, and CAMPD file (json or csv) download.
1. account-mgmt
   - allowance-holdings
   - accounts/attributes
   - allowance-compliance
   - emissions-compliance
2.  facilities-mgmt
   - facilities/attributes
3. streaming-services
   - facilities/attributes
   - accounts/attributes
   - allowance-holdings 
   - allowance-compliance
   - emissions-compliance

buyOwner & sellOwner is found in the following APIs/endpoints and should have the | delimiter applied between each owner/operator value in a record e.g., "Georgia Power Company (Owner)|Southern Company (Owner)|Alabama Power Company (Operator)" which should be returned in the API call, CAMPD UI, and CAMPD file (json or csv) download.
1. account-mgmt
   - allowance-transactions
2. streaming-services
   - allowance-transactions
 
*Note* in the account-mgmt API there are endpoints for each data type that end in /applicable or /owner-operators but I am not sure if these are effected by this work. I think not, but wanted to make sure they were known.

The above requirements should not interfere with the current functionality of the CAMPD UI and APIs where a user can input one or more owner/operator(s) and filter for records matching the inputted owner/operator(s). The filters currently exist for the:
- account-mgmt endpoints listed above
- streaming services endpoints listed above (except facilities/attributes)

@alangmaid

**Comment by alangmaid** [comment]
On the UI, the filters that include "Owner", Operator", or "Owner/Operator" that need to be verified as having the correct "|" treatment are (current verification status on staging in UI and download):

- Allowance, Account Information (correct)
- Allowance, Holdings (correct)
- Allowance, Transactions (correct)
- Compliance, Alllowance Based (correct)
- Compliance, Emissions Based (incorrect - Owner/ Operator displays "," instead of "|" between owner and operator. "|" displays as expected between multiple owners and multiple operators.)
- Facility, Facility/Unit Attributes (incorrect - Owner/ Operator displays "," instead of "|" between owner and operator. "|" displays as expected between multiple owners and multiple operators.)

**Comment by alangmaid** [comment]
Retagging this ticket to v1.4 to address remaining items which are to address the concatenation that occurs on the "Owner/Operator" columns in:
Compliance, Emissions Based (Owner/ Operator displays "," instead of "|" between owner and operator. "|" displays as expected between multiple owners and multiple operators.)
Facility, Facility/Unit Attributes (Owner/ Operator displays "," instead of "|" between owner and operator. "|" displays as expected between multiple owners and multiple operators.)

**Comment by lgiannini1** [comment]
@shahbazkhan1999 On the Compliance, Emissions Based and Facility, Facility/Unit Attributes Custom Data Downloads, Owner' Operator still displays a "," instead of "|" between owner and operator in the data preview and CSV files. 

Compliance, Emissions Based:
<img width="1339" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOVpHQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--3977e74b59e53874a70874339844783a99f93dfe/image.png" alt="image.png" />

Compliance, Emissions Based CSV Download:
<img width="1033" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBd0ZIQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--549535f356667d3da3caf8f338571f8340840c01/image.png" alt="image.png" />

Facility, Facility/ Unit Attributes:
<img width="925" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBd2RIQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--5aa9b6c8fd52b468e0c14e72a6e2f3aeb939d316/image.png" alt="image.png" />

Facility CSV Download:
<img width="853" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBd3BIQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--f77672da77cbf378a3ed67a58706e0fbda65b1e6/image.png" alt="image.png" />

**Comment by shahbazkhan1999** [comment]
Changes are already been made; I am just waiting for the PR to approve so create a new release and added @djw4erg as a reviewer

**Comment by lgiannini1** [comment]
Blocked by #6101

**Comment by ergjustin** [comment]
@lgiannini1 
#6101 I revert to previous (disable CSP). I think you can test from now onwards.

**Comment by lgiannini1** [comment]
@shahbazkhan1999 I am still seeing the issues noted above where a "," is displayed instead of "|" between owners and operators in the Compliance, Emissions Based and Facility, Facility/Unit Attributes Data Previews. The delimiter is displayed correctly in the Compliance, Emissions Based CSV Download, but not in the Facility, Facility/ Unit Attributes download

Compliance, Emissions Based Data Preview:
![image](https://github.com/US-EPA-CAMD/easey-ui/assets/63416419/287052ab-9eec-4592-bb77-d0d1cb36486a)

Facility, Facility/Unit Attributes Data Preview:
![image](https://github.com/US-EPA-CAMD/easey-ui/assets/63416419/c10007de-38bc-4d43-885f-e0b165405819)

Facility, Facility/ Unit Attributes CSV Download:
![image](https://github.com/US-EPA-CAMD/easey-ui/assets/63416419/9a7cbed0-e0a1-45ac-a774-627989051f78)

**Comment by shahbazkhan1999** [comment]
@lgiannini1 changes were merged but not deployed I do not have the access for that but now they are deployed you can test.


![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/136729393/55e02658-8ae9-42c7-bd06-0748acd42dfb)

**Comment by lgiannini1** [comment]
Delimiters look correct in Facility/Unit Attributes, and all Allowance and Compliance data sub types

**Comment by szintgraff** [comment]
Checked Owner/Operator Column in Facility, Allowance (Account Info, Holdings, and Transactions), and Compliance (Emissions and Allowance based) on Data Queries in staging: https://campd-stg.app.cloud.gov/data/custom-data-download. No comments on the facility, allowance, or allowance based compliance data.

For compliance data, emissions based: The Owner/Operator Column doesn't have a delimiter and seems to only have operator data in that column. Is that how the data is suppose to display? It is also that way in the production environment.

@lgiannini1

**Comment by szintgraff** [comment]
Moving to ready for deployment- Checked with Travis and the owner/operator column looks correct in the emissions based compliance.

**Comment by ergjustin** [comment]
Issue: 1
Data Type: Compliance
Data Subtype: Allowance Based
programCode : ARP
facilityId: 2828

![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/108369391/f3a6295e-1849-4740-80ce-56bb030989c8)



Issue : 2
The `ownerOperator` examples on the Swagger page use a comma instead of a pipe

**Comment by ergjustin** [comment]
Made changes in 

1.  `easey-account-api`: Removed extra pipe in `ownerOperator` and The `ownerOperator` examples on the Swagger page
2. `easey-facilities-api`: The ownerOperator examples on the Swagger page
3.  `easey-streaming-services` : The ownerOperator examples on the Swagger page

**Comment by ergjustin** [comment]
Developer test
1. easey-account-api

![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/108369391/c29a5594-6c99-4067-97d1-796f197a8729)
![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/108369391/87b2912c-d1d7-4efe-8666-5260490eb26e)
2. easey-facilities-api
![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/108369391/e507a9f1-9ec4-43e7-ae9a-0eae59a512ff)
3. easey-streaming-services
![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/108369391/f5cc3e8f-e2dc-42aa-b3f5-20944fdcc90a)

**Comment by lgiannini1** [comment]
In the Facility, Facility/Unit Attributes CSV download, a "," is displayed instead of "|" between owner and operator.

![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/63416419/1362b499-9edc-4ac7-8a33-1d936e7ebc0c)

On the streaming services swagger page, a "," is sometimes displayed instead of "|" between owners and operators

Result when using year=2021 and stateCode=IA:
![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/63416419/ba714567-f6c5-45ef-84aa-b39629888684)

Results when using year=2021 and facilityId=2828:
![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/63416419/986ea2f4-5fe6-42f6-8680-4963cc9327bd)

![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/63416419/d62156e4-b949-4a17-af0c-5a10863866a4)

**Comment by ergjustin** [comment]
Fixed the above issue 
Developer test.
![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/108369391/bce1bce6-6ba1-4b8e-92ce-aedaab66c2dd)
![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/108369391/e2caebb4-05ab-4e92-925e-17a7944d359b)

**Comment by lgiannini1** [comment]
Delimiters look correct on Facility/Unit Attributes, and all Allowance and Compliance data sub types in UI, CSV Downloads, and Swagger pages.

**Comment by tlageman-epa** [comment]
Took a look at this and agree with @lgiannini1 on each part except swagger pages. The data return is fine, but the example schemas need some tweaking to accurately reflect the delimiter patter of "value|value". Right now, some of them have "value , value" and "value | value". Here the corrections needed.

Account Mgmt
	- Holdings - swagger page example schema - extra space around |
	- Transactions - swagger page example schema - has , instead of |
	- Attributes - swagger page example schema - extra space around |
	- Account-based compliance - swagger page example schema - extra space around |
	- Emissions-based compliance - swagger page example schema - extra space around |

Facilities Mgmt
	- Attributes - swagger page example schema - extra space around |

Streaming Services
	- Facility attributes - swagger page example schema - extra space around |
	- Account attributes - swagger page example schema - extra space around |
	- Holdings - swagger page example schema - extra space around |
	- Transactions - swagger page example schema - has , instead of |
	- Account-based compliance - swagger page example schema - extra space around |
Emissions-based compliance - swagger page example schema - extra space around |

FYI @szintgraff

**Comment by ergjustin** [comment]
@tlageman-epa 
For the clarification : Transactions - swagger page example schema - has , instead of |  Is it in the  `sellOwner` ?

**Comment by tlageman-epa** [comment]
@ergjustin yes, buyOwner is fine

**Comment by ergjustin** [comment]
Developer Test
Fixed
1. swagger page example schema - extra space around |

![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/108369391/d51f4f59-a4cc-4665-9358-0c7f911477b2)
2.  swagger page example schema - has , instead of |
![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/108369391/1e3feb11-6b53-463d-bc3e-b6e39c64a5fa)

**Comment by lgiannini1** [comment]
Example schemas on account-mgmt, facilities-mgmt, and streaming-services swagger pages have the correct delimiters for ownerOperator with no additional spaces.

**Comment by szintgraff** [comment]
@spetros-do  

I checked facility, allowance account (account info, transaction, holding) and compliance (allowance and emissions based). I found one issue outlined below.

There is a pipe where I think a coma should be for "Louisiana 1". 
To find the screen shot below in facility data preview, I set the year = 2019, and facility = Louisiana 1. There may be other instances for that facility for example, I see it in 2020 and 2021 too. There are at least 14 instances of this incorrect pipe. 


![Image](https://github.com/US-EPA-CAMD/easey-ui/assets/115799210/a34fcf8d-c42e-45a3-b788-1ef54bf36b61)

**Comment by ergjustin** [comment]
Till now owner/operator was split with “),” then it joins with “)|”.
When the above situation “),” splits an extra pipe joins. The reason for that is owner/operator has (West),LLC which joins the pipe. 

Solution: 
05/09
```const splitOwnWithPipe = entity.ownDisplay?.split('|');
const splitOprWithPipe = entity.oprDisplay?.split('|');
const uniqueOwn = [...new Set(splitOwnWithPipe)].join('|');
const uniqueOpr = [...new Set(splitOprWithPipe)].join('|');
const uniqueOwnOprList = [uniqueOwn, uniqueOpr];
const ownerOperator = uniqueOwnOprList.filter(e => e).join('|');

**Comment by szintgraff** [comment]
@mark-hayward-erg @ergjustin @alangmaid 

This explanation doesn't align with what I understand about the approach for this ticket. Let's talk about it during the CAMPD weekly meeting later today. My questions are:

- Which database is this data coming from?
- What does it look like in the original database?
- Are there multiple operators in one column?
- If not, why are we splitting by ")," and then rejoining?

**Comment by lgiannini1** [comment]
Delimiters look correct on Facility/Unit Attributes, and all Allowance and Compliance data sub types in UI, CSV Downloads, and Swagger pages for ORIS 1391.

**Comment by szintgraff** [comment]
I tested the locations below and am moving to ready for deployment. 
- Downloaded facility data from 2015-2021 
- Compliance data allowance based for 2021 and 2022
- Compliance emission based data from 2015-2021
- Allowance, Account Info for all states
- Allowance, holding from 2015-2021
- Allowance transaction data from 2020-2021

Testing description:
I spot checked unique occurrences of owner/operator columns in the following data pulls and specifically looked for "space|", "|space", "| INC", "| LLC", "| LP", "| LTD".
